### PR TITLE
core: ree_fs: check ta size early

### PR DESCRIPTION
**Pull request by jenswi-linaro**
As soon as TA signature has been verified check that the total loaded size matches what is in the sign header. This prevents an eventual attacker from providing arbitrary values in the img_size field of the signed header.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

File: core/kernel/ree_fs_ta.c [hunk 1]
```diff
@@ -229,6 +229,31 @@ static TEE_Result rpc_load(const TEE_UUID *uuid, struct shdr **ta,
...
+		break;
+	case SHDR_ENCRYPTED_TA:
+		hdr_sz = sizeof(struct shdr_bootstrap_ta) +
+			 sizeof(struct shdr_encrypted_ta);
+		break;
+	default:
+		return TEE_ERROR_SECURITY;
+	}
+
+	if (ta_size != offs + hdr_sz + shdr->img_size)
```
Comment (etienne-lms) [inline core/kernel/ree_fs_ta.c:251]: we should add the signed heade size also I think (from `SHDR_GET_SIZE(shdr)` or like).

Sometthing like:
```c
static TEE_Result check_ta_type_and_size(const struct shdr *shdr, size_t offs,
					 size_t ta_size, uint8_t *ta)
{
	struct shdr_encrypted_ta img_ehdr = { };
	size_t hdr_sz = SHDR_GET_SIZE(shdr);

	switch (shdr->img_type) {
	case SHDR_TA:
		break;
	case SHDR_BOOTSTRAP_TA:
		hdr_sz += sizeof(struct shdr_bootstrap_ta);
		break;
	case SHDR_ENCRYPTED_TA:
		hdr_sz += sizeof(struct shdr_bootstrap_ta);
		memcpy(&img_ehdr, ta + offs + hdr_sz, sizeof(img_ehdr));
		hdr_sz += SHDR_ENC_GET_SIZE(&img_ehdr);
		break;
	default:
		return TEE_ERROR_SECURITY;
	}

	if (ta_size != offs + hdr_sz + shdr->img_size)
		return TEE_ERROR_SECURITY;

	return TEE_SUCCESS;
}
```
Maybe some ADD_OVERFLOW are missing.

File: core/kernel/ree_fs_ta.c [hunk 2]
```diff
@@ -229,6 +229,31 @@ static TEE_Result rpc_load(const TEE_UUID *uuid, struct shdr **ta,
...
+
+	switch (shdr->img_type) {
+	case SHDR_TA:
+		break;
+	case SHDR_BOOTSTRAP_TA:
+		hdr_sz = sizeof(struct shdr_bootstrap_ta);
+		break;
+	case SHDR_ENCRYPTED_TA:
+		hdr_sz = sizeof(struct shdr_bootstrap_ta) +
+			 sizeof(struct shdr_encrypted_ta);
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/kernel/ree_fs_ta.c:245]: You're right.

File: core/kernel/ree_fs_ta.c (hunk 1 above)
Comment (jenswi-linaro) [reply core/kernel/ree_fs_ta.c:251]: The initial `SHDR_GET_SIZE(shdr)` isn't needed, that's already added to offs.

Comment (etienne-lms) [reply core/kernel/ree_fs_ta.c:251]: It was added only for the subkey headers found in the binary, not for the TA image (bootstrap, encrypted or installed) part were looking into here.

**Comment by jbech-linaro** [comment]
I'd suggest to add @asafmod to one or both to the commit message(s), i.e.,
```
Suggested-by: Asaf Modelevsky <amodele@amazon.com>
Reported-by: Asaf Modelevsky <amodele@amazon.com>
```

**Comment by jenswi-linaro** [comment]
This became more complicated than anticipated. I'm going to start over.

**Comment by jenswi-linaro** [comment]
Updated with a new approach. I found another small issue that I'm addressing in the last patch.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by jenswi-linaro** [comment]
Tag applied

**Pull request merged**